### PR TITLE
Added converters to handle the v2.1.0 collection schema

### PR DESCRIPTION
**Pull request by kunagpal**
In addition to new code to handle conversions to and from v2.1.0, this pull request also re-arranges the existing converter structure, making it a little easier to add the capability to handle more versions in the future.
 
Some of this code has been repurposed from https://github.com/postmanlabs/postman-collection-transformer/compare/feature/auth-params-as-array.

### CODE REVIEW

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            if (_.has(newItem, 'id')) {
+                !_.has(newItem, '_postman_id') && (newItem._postman_id = newItem.id);
+                delete newItem.id;
+            }
+
+            newItem.request = self.request(newItem && newItem.request);
+            newItem.response = _.map(newItem.response || newItem.responses, function (response) {
+                return self.response(response);
+            });
+
```
Comment (shamasis) [inline lib/converters/v2.0.0/converter-v2-to-v21.js]: where do we handle saved responses?

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    convert: function (collection, options, callback) {
+        var units = ['item'],
+            builders = new Builders(options),
+            info = collection && collection.info,
+            newCollection = {
+                info: {
+                    _postman_id: builders.options.retainIds ? info && (info._postman_id || info.id) : undefined
+                },
+                // this is done to avoid mutating the original collection
+                variables: (options && options.env) ? builders.variables(options.env) : []
```
Comment (shamasis) [inline lib/converters/v2.0.0/converter-v2-to-v21.js]: `variables` key is defined even if none defined?

File: lib/converters/v2.1.0/converter-v21-to-v1.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+     * @param {Boolean} skipResponses - Will result in saved responses being ignored for conversion if set to true.
+     * @returns {Object} An equivalent v1.0.0 representation for the current v2.0.0.1 request.
+     */
+    request: function (item, collectionId, skipResponses) {
+        var _item = _.cloneDeep(item), // do not mutate the original
+            auth = _.get(_item, 'request.auth');
+
+        if (auth && auth.type && _.isArray(auth[auth.type])) {
+            auth[auth.type] = _.transform(auth[auth.type], function (a, element) {
+                a[element.key] = element.value;
```
Comment (shamasis) [inline lib/converters/v2.1.0/converter-v21-to-v1.js]: we don't do type conversion or force string here?

File: lib/converters/v1.0.0/converter-v1-to-v21.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+     * @returns {{type: *}} - The v2.1.0 compliant request object
+     */
+    auth: function (requestV1) {
+        var auth = Builders.super_.prototype.auth.call(this, requestV1);
+
+        if (!(auth && auth.type && auth[auth.type])) { return; }
+
+        return {
+            type: auth.type,
+            [auth.type]: _.map(auth[auth.type], function (value, key) {
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: To be changed to ES5.

File: lib/converters/v1.0.0/converter-v1-to-v21.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                return { key: key, value: value };
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: This is a conversion from an object to a list of key value pairs, so there is no type information to extract.

File: lib/converters/v1.0.0/converter-v1-to-v21.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+     * @param {Object} collection - The V1 collection instance to convert to a v2.1.0 format.
+     * @param {Object} options - The set of options for the current conversion sequence.
+     * @param {?Function} callback - The function invoked to mark the completion of the conversion process/
+     * @returns {*}
+     */
+    convert: function (collection, options, callback) {
+        var units = ['info', 'item'],
+            builders = new Builders(options),
+            newCollection = {
+                variables: (options && options.env) ? builders.variables(options.env) : []
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: Yes, this is the behaviour all across.

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        return newRes;
+    },
+
+    /**
+     * Normalizes incoming itemGroups (folders) for v2.1.0 format compliance.
+     *
+     * @param {Object} itemGroup - The folder to be converted to the v2.1.0 format.
+     * @returns {Object} - The processed item group in v2.1.0 format.
+     */
+    itemGroup: function (itemGroup) {
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Yes, to be done in the next P.R 😄

File: lib/converters/v2.0.0/converter-v2-to-v21.js (hunk 2 above)
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Yes.

File: lib/converters/v2.1.0/converter-v21-to-v1.js (hunk 3 above)
Comment (kunagpal) [reply lib/converters/v2.1.0/converter-v21-to-v1.js]: No, some parameters can be boolean as well. (Those are primarily UI based)

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    v2Common = require('../common/v2'),
+    constants = require('../../constants'),
+
+    Builders = function ConversionBuilder (options) {
+        this.options = options || {};
+    };
+
+_.assign(Builders.prototype, {
+
+    url: function (url) {
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: That is for conversion between strings and url object representations, this is to ensure that the `id` in path variables is always renamed to `key`, as discussed earlier.

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        req && req.url && (req.url = self.url(req.url));
+
+        if (auth) {
+            // the auth structure is the only difference across the v2.0.0 and v2.1.0 formats
+            // conversion is done as below to handle cases where a single request may have multiple auth manifests
+            _.forEach(['aws', 'basic', 'digest', 'hawk', 'noauth', 'oauth1', 'oauth2'], function (type) {
+                // only convert if auth params aren't already an array.
+                if (_.has(auth, type) && !_.isArray(auth[type])) {
+                    auth[type] = _.map(auth[type], function (value, key) {
+                        return { key: key, value: value };
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Similar to the comment above.

File: lib/converters/v2.0.0/converter-v2-to-v21.js (hunk 1 above)
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Here 😓

File: lib/converters/v2.0.0/converter-v2-to-v1.js [hunk 10]
```diff
@@ -572,25 +539,22 @@ module.exports = {
             _.set(collection, 'info.id', util.uid());
         }
 
-        var units = ['order', 'folders_order', 'folders', 'requests'],
+        var tempCollection = _.cloneDeep(collection),
```
Comment (kamalaknn) [inline lib/converters/v2.0.0/converter-v2-to-v1.js]: @kunagpal I can see your intention to not mutate the input collection. Isn't this expensive?

File: lib/converters/v2.1.0/converter-v21-to-v2.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+     * @returns {Object} An equivalent v2.0.0 representation for the current v2.1.0 request.
+     */
+    request: function (request) {
+        var req = _.cloneDeep(request), // do not mutate the original
+            auth = req && req.auth;
+
+        if (auth) {
+            // the auth structure is the only difference across the v2.1.0 and v2.0.0 formats
+            // conversion is done as below to handle cases where a single request may have multiple auth manifests
+            _.forEach(['aws', 'basic', 'digest', 'hawk', 'noauth', 'oauth1', 'oauth2'], function (type) {
```
Comment (kamalaknn) [inline lib/converters/v2.1.0/converter-v21-to-v2.js]: @kunagpal here as well

File: lib/converters/v2.0.0/converter-v2-to-v1.js (hunk 10 above)
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v1.js]: It is, but the transformer is not meant to mutate any incoming data, something that can be seen from the existing converters.

File: lib/converters/v2.0.0/converter-v2-to-v21.js [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        var self = this,
+            req = _.cloneDeep(request), // do not mutate the original
+            auth = req && req.auth;
+
+        req && req.url && (req.url = self.url(req.url));
+
+        if (auth) {
+            // the auth structure is the only difference across the v2.0.0 and v2.1.0 formats
+            // conversion is done as below to handle cases where a single request may have multiple auth manifests
+            _.forEach(['aws', 'basic', 'digest', 'hawk', 'noauth', 'oauth1', 'oauth2'], function (type) {
```
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Yes, this is meant to be removed. Good catch 👍

[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: Removed.

File: lib/converters/v2.1.0/converter-v21-to-v2.js (hunk 11 above)
Comment (kunagpal) [reply lib/converters/v2.1.0/converter-v21-to-v2.js]: Removed.

File: lib/converters/v1.0.0/converter-v1-to-v21.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: Done.

File: lib/converters/v1.0.0/converter-v1-to-v21.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (shamasis) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: Can we not change that for 2.1? Seems redundant.

File: lib/converters/v2.0.0/converter-v2-to-v21.js (hunk 2 above)
Comment (shamasis) [reply lib/converters/v2.0.0/converter-v2-to-v21.js]: let's change this for 2.1? any downside to that?

File: lib/converters/v2.0.0/converter-v2-to-v1.js (hunk 10 above)
Comment (shamasis) [reply lib/converters/v2.0.0/converter-v2-to-v1.js]: ok

File: lib/converters/v2.1.0/converter-v21-to-v1.js (hunk 3 above)
Comment (shamasis) [reply lib/converters/v2.1.0/converter-v21-to-v1.js]: question is when we do v1 to v2, do we use the original type to set type?

**Pull request merged**